src/Symfony/Component/Messenger/Middleware/AddBusNameStampMiddleware.php

```diff
@@ -32,7 +32,9 @@ public function __construct(string $busName)
 
     public function handle(Envelope $envelope, StackInterface $stack): Envelope
     {
-        $envelope = $envelope->with(new BusNameStamp($this->busName));
+        if (null === $envelope->last(BusNameStamp::class)) {
+            $envelope = $envelope->with(new BusNameStamp($this->busName));
+        }
```

Comment:
Fixes an unrelated "bug" where retries would cause many BusNameStamp to be added, which was just unnecessary and wasteful. Test was added for this.